File: cabot/cabot/event.py
# Copyright (c) 2020, 2022  Carnegie Mellon University
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

class BaseEvent(object):
    def __init__(self, type=None):
        self._type = type

    @property
    def type(self):
        return self._type

    def __str__(self):
        raise RuntimeError(F"event({self.type})")

    @classmethod
    def parse(cls, text):
        for c in cls.__subclasses__():
            # check the class has own _parse
            inst = c._parse(text)
            if inst is not None:
                return inst

            # otherwise check subclass parse
            inst = c.parse(text)
            if inst is not None:
                return inst

        return None


class ButtonEvent(BaseEvent):
    TYPE = "button"

    def __init__(self, type=None, button=0, up=False, hold=False):
        type = type if type is not None else ButtonEvent.TYPE
        super(ButtonEvent, self).__init__(type=type)
        self._button = button
        self._up = up
        self._hold = hold

    def __eq__(self, other):
        return self.type == other.type \
            and self.button == other.button \
            and self.up == other.up \
            and self.hold == other.hold

    @property
    def button(self):
        return self._button

    @property
    def up(self):
        if self._hold:
            return False
        return self._up

    @property
    def down(self):
        if self._hold:
            return False
        return not self._up

    @property
    def hold(self):
        return self._hold

    def __str__(self):
        subtype = "hold" if self.hold else "up" if self.up else "down"
        return F"{self.type}_{subtype}_{self.button}"

    @classmethod
    def _parse(cls, text):
        if text.startswith(cls.TYPE):
            items = text.split("_")
            hold = False
            up = False
            if items[1] == "hold":
                hold = True
            if items[1] == "up":
                up = True
            button = int(items[2])
            return cls(button=button, up=up, hold=hold)
        return None


class JoyButtonEvent(ButtonEvent):
    TYPE = "joybutton"

    def __init__(self, button=0, up=False, hold=False):
        super(JoyButtonEvent, self).__init__(type=JoyButtonEvent.TYPE,
                                             button=button, up=up, hold=hold)


class ClickEvent(BaseEvent):
    TYPE = "click"

    def __init__(self, type=None, buttons=0, count=0):
        type = type if type is not None else ClickEvent.TYPE
        super(ClickEvent, self).__init__(type=type)
        self._buttons = buttons
        self._count = count

    def __eq__(self, other):
        return self.type == other.type \
            and self.buttons == other.buttons \
            and self.count == other.count

    @property
    def buttons(self):
        return self._buttons

    @property
    def count(self):
        return self._count

    def __str__(self):
        return F"{self.type}_{self.buttons}_{self.count}"

    @classmethod
    def _parse(cls, text):
        if text.startswith(cls.TYPE):
            items = text.split("_")
            buttons = int(items[1])
            count = int(items[2])
            return cls(buttons=buttons, count=count)
        return None


class HoldDownEvent(BaseEvent):
    TYPE="holddown"
    def __init__(self, type=None, holddown=0):
        type = type if type is not None else HoldDownEvent.TYPE
        super(HoldDownEvent, self).__init__(type=type)
        self._holddown = holddown

    def __eq__(self, other):
        return self.type == other.type \
            and self.holddown == other.holddown

    @property
    def holddown(self):
        return self._holddown

    def __str__(self):
        return "%s_%d" % (self.type, self.holddown)

    @classmethod
    def _parse(cls, text):
        if text.startswith(cls.TYPE):
            items = text.split("_")
            holddown = int(items[1])
            return cls(holddown=holddown)
        return None


class JoyClickEvent(ClickEvent):
    TYPE = "joyclick"

    def __init__(self, buttons=0, count=0):
        super(JoyClickEvent, self).__init__(type=JoyClickEvent.TYPE,
                                            buttons=buttons, count=count)

File: cabot/src/cabot/event.cpp
/*******************************************************************************
 * Copyright (c) 2023  Miraikan and Carnegie Mellon University
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *******************************************************************************/

#include "event.hpp"
#include <string>
#include <memory>
#include <vector>

BaseEvent::BaseEvent(std::string type)
: _type(type) {}
std::string BaseEvent::getType() const
{
  return _type;
}
std::string BaseEvent::toString() const
{
  throw std::runtime_error("event(" + _type + ")");
}
BaseEvent * BaseEvent::parse(const std::string & text)
{
  for (const std::unique_ptr<BaseEvent> & subclass : getSubclasses()) {
    BaseEvent * inst = subclass->_parse(text);
    if (inst != nullptr) {
      return inst;
    }
    inst = subclass->parse(text);
    if (inst != nullptr) {
      return inst;
    }
  }
  return nullptr;
}

BaseEvent * BaseEvent::_parse(const std::string&, const std::string&)
{
  return nullptr;
}

std::vector<std::unique_ptr<BaseEvent>> & BaseEvent::getSubclasses()
{
  static std::vector<std::unique_ptr<BaseEvent>> subclasses;
  return subclasses;
}

ButtonEvent::ButtonEvent(int button, bool up, bool hold)
: BaseEvent("button"), _button(button), _up(up), _hold(hold) {}
bool ButtonEvent::operator==(const ButtonEvent & other) const
{
  return getType() == other.getType() && _button == other._button && _up == other._up &&
         _hold == other._hold;
}
int ButtonEvent::get_button() const
{
  return _button;
}
bool ButtonEvent::is_up() const
{
  return !_hold && _up;
}
bool ButtonEvent::is_down() const
{
  return !_hold && !_up;
}
bool ButtonEvent::is_hold() const
{
  return _hold;
}
std::string ButtonEvent::toString() const
{
  std::string subtype = is_hold() ? "hold" : (is_up() ? "up" : "down");
  return getType() + "_" + subtype + "_" + std::to_string(_button);
}
BaseEvent * ButtonEvent::_parse(const std::string & text, const std::string&)
{
  if (text.find(TYPE) == 0) {
    std::vector<std::string> items;
    boost::split(items, text, boost::is_any_of("_"));
    bool hold = false;
    bool up = false;
    if (items[1] == "hold") {
      hold = true;
    }
    if (items[1] == "up") {
      up = true;
    }
    int button = std::stoi(items[2]);
    return new ButtonEvent(button, up, hold);
  }
  return nullptr;
}

const char* ButtonEvent::TYPE = "button";

JoyButtonEvent::JoyButtonEvent(int button, bool up, bool hold)
: ButtonEvent(button, up, hold) {}

const char* JoyButtonEvent::TYPE = "joybutton";

ClickEvent::ClickEvent(int buttons, int count)
: BaseEvent("click"), _buttons(buttons), _count(count) {}
bool ClickEvent::operator==(const ClickEvent & other) const
{
  return getType() == other.getType() && _buttons == other._buttons && _count == other._count;
}
int ClickEvent::get_buttons() const
{
  return _buttons;
}
int ClickEvent::get_count() const
{
  return _count;
}
std::string ClickEvent::toString() const
{
  return getType() + "_" + std::to_string(_buttons) + "_" + std::to_string(_count);
}
BaseEvent * ClickEvent::_parse(const std::string & text, const std::string&)
{
  if (text.find(TYPE) == 0) {
    std::vector<std::string> items;
    boost::split(items, text, boost::is_any_of("_"));
    int buttons = std::stoi(items[1]);
    int count = std::stoi(items[2]);
    return new ClickEvent(buttons, count);
  }
  return nullptr;
}

const char* ClickEvent::TYPE = "click";

HoldDownEvent::HoldDownEvent(int holddown)
: BaseEvent("holddown"), _holddown(holddown) {}
bool HoldDownEvent::operator==(const HoldDownEvent & other) const
{
  return getType() == other.getType() && _holddown == other._holddown;
}
int HoldDownEvent::get_holddown() const
{
  return _holddown;
}
std::string HoldDownEvent::toString() const
{
  return getType() + "_" + std::to_string(_holddown);
}

const char* HoldDownEvent::TYPE = "holddown";

BaseEvent * HoldDownEvent::_parse(const std::string & text, const std::string&)
{
  if (text.find(TYPE) == 0) {
    std::vector<std::string> items;
    boost::split(items, text, boost::is_any_of("_"));
    int holddown = std::stoi(items[1]);
    return new HoldDownEvent(holddown);
  }
  return nullptr;
}

JoyClickEvent::JoyClickEvent(int buttons, int count)
: ClickEvent(buttons, count) {}

const char* JoyClickEvent::TYPE = "joyclick";
